stl/inc/flat_map

```diff
@@ -741,34 +737,79 @@ protected:
-        auto _Key_it = _STD lower_bound(_Data.keys.begin(), _Data.keys.end(), _Key_val, _Key_compare);
-        if (_Key_it != _Data.keys.end() && _Key_equal(*_Key_it, _STD forward<_KeyTy>(_Key_val))) {
-            // Already exists
-            return {this->begin() + _STD distance(_Data.keys.begin(), _Key_it), false};
+    template <bool _OverwriteIfExists, class _OtherKey, class... _MappedArgTypes>
+    iterator _Emplace_hint(const_iterator _Position, _OtherKey&& _Key_val, _MappedArgTypes&&... _Args) {
+        _STL_INTERNAL_STATIC_ASSERT(
+            is_constructible_v<mapped_type, _MappedArgTypes...>
+            && (is_same_v<remove_cvref_t<_OtherKey>, key_type>
+                || (is_constructible_v<key_type, _OtherKey> && _Is_transparent_v<key_compare>) ));
+        static_assert(_IsUnique || !_OverwriteIfExists,
+            "Overwriting is not supported when the container allows multiple copies of a key.");
+        const const_iterator _Begin = cbegin();
+        const const_iterator _End   = cend();
+
+        bool _Insert_before_position        = false;
+        bool _Insert_after_position_minus_1 = false;
+        if constexpr (_IsUnique) {
+            _Insert_before_position        = (_Position == _End) || _Key_compare(_Key_val, *(_Position._Key_it));
+            _Insert_after_position_minus_1 = (_Position == _Begin) || _Key_compare(*(_Position._Key_it - 1), _Key_val);
         } else {
-            // Need to insert
-            auto _Index = _STD distance(_Data.keys.begin(), _Key_it);
-            this->_Insert_exact(this->cbegin() + _Index, key_type{_STD forward<_KeyTy>(_Key_val)},
-                mapped_type{_STD forward<_MappedArgTypes>(_Mapped_args)...});
-            return {this->begin() + _Index, true};
+            _Insert_before_position        = (_Position == _End) || !_Key_compare(*(_Position._Key_it), _Key_val);
+            _Insert_after_position_minus_1 = (_Position == _Begin) || !_Key_compare(_Key_val, *(_Position._Key_it - 1));
+        }
+        bool _Hint_is_accurate = _Insert_before_position && _Insert_after_position_minus_1;
+
+        if (_Hint_is_accurate) {
+            auto _Dist = _STD distance(_Begin._Key_it, _Position._Key_it);
+            {
+                key_type _Key_to_insert(_STD forward<_OtherKey>(_Key_val));
+                mapped_type _Mapped_to_insert(_STD forward<_MappedArgTypes>(_Args)...);
+                _Insert_exact(_Position, _STD move(_Key_to_insert), _STD move(_Mapped_to_insert));
+            }
+            return begin() + _Dist;
+        } else {
+            if constexpr (_OverwriteIfExists) {
+                if (_Key_equal(_Key_val, *(_Position._Key_it))) {
+                    auto _Dist        = _STD distance(_Begin._Key_it, _Position._Key_it);
+                    auto _It          = begin() + _Dist;
+                    *(_It._Mapped_it) = mapped_type{_STD forward<_MappedArgTypes>(_Args)...};

```

Comment:
Question: As much as I love braces for constructing temporaries, are we allowed to use them here? What if we're trying to construct a `vector<int>` as the mapped type from `5` copies of `1729`? Occurs below. May indicate missing test coverage.

(When we construct `_Mapped_to_insert`, we use parentheses, so I strongly suspect braces are a bug.)